homeassistant/components/sensor/uscis.py

```diff
@@ -0,0 +1,116 @@
+"""
+Support for USCIS Case Status.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.uscis/
+"""
+
+import logging
+from datetime import timedelta
+import voluptuous as vol
+
+from homeassistant.helpers.entity import Entity
+from homeassistant.util import Throttle
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.helpers import config_validation as cv
+from homeassistant.const import CONF_FRIENDLY_NAME
+
+
+_LOGGER = logging.getLogger(__name__)
+
+REQUIREMENTS = ['lxml==3.5.0']
+
+DEFAULT_NAME = "USCIS"
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_FRIENDLY_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Required('case_id'): cv.string,
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setting the platform in HASS and Case Information."""
+    uscis = UscisSensor(config['case_id'], config[CONF_FRIENDLY_NAME])
+    uscis.update()
+    if uscis.valid_case_id:
+        add_devices([uscis])
+    else:
+        _LOGGER.error("Setup USCIS Sensor Fail"
+                      " check if your Case ID is Valid")
+    return
+
+
+class UscisSensor(Entity):
+    """USCIS Sensor will check case status on daily basis."""
+
+    HOURS_TO_UPDATE = timedelta(hours=24)
+
+    CURRENT_STATUS = "current_status"
+    LAST_CASE_UPDATE = "last_update_date"
+    CASE_DATE_PATTERN = r"[(A-Za-z)]*\s[\d]*,\s[\d]*"
+    URL = "https://egov.uscis.gov/casestatus/mycasestatus.do"
+    APP_RECEIPT_NUM = "appReceiptNum"
+    INIT_CASE_SEARCH = "initCaseSearch"
+    CASE_STATUS = "CHECK STATUS"
+    UPDATE_TEXT_XPATH = "/html/body/div[2]/form/div/div[1]" \
```

Comment:
This is too much protocol specific information and should be extracted into an external package.